### ISSUE TITLE
Autogenerate ROS2 API Documentation on Buildfarm

### ISSUE THREAD

**Issue by vatanaksoytezer** [osrf/infrastructure#10]
- [x] I have checked [ROS Answers](https://answers.ros.org) for similar reports and solutions.
- [x] I have checked the [ROS Status page](https://status.ros.org) and my issue is not listed there.
- [x] I have checked the [open issues](https://github.com/osrf/infrastructure/issues) for similar reports.
- [x] This is **not** a [security issue](https://github.com/osrf/infrastructure/blob/latest/SECURITY.md) If you have a security issue to report please do so via email to <security@openrobotics.org>.

## Problem description

We would like to publish our doxygen generated documentation to [ROS2 API page](https://docs.ros2.org/foxy/api/). I've searched a bit and hit [this post](https://answers.ros.org/question/319362/how-to-build-documentation-for-ros2-packages/) from 2019. Is there any updates on that to make buildfarm auto (or even manually) generate and publish the documentation to ROS2 API page? I see some documentation already published there from Open Robotics repositories. Is there a new way of making the buildfarm to publish these, or are these manually published?

**Comment by vatanaksoytezer** [comment]
Friendly ping to some people that may have insights on this: @clalancette @sloretz

**Comment by tylerjw** [comment]
Just adding another ping to this.  We would really like to use the buildfarm to generate API documentation as it did before in ROS 1.

**Comment by clalancette** [comment]
So the short of it is that we have a bunch of the infrastructure in place to make this happen, but not all of it.  You can see what we have at https://build.ros2.org/view/Rdoc/ , which are the jobs that generate the documentation.

There's a few things missing that are important to make it fully work:
1. We need to create Rdoc_reconfigure jobs and Rdoc_trigger jobs so that the jobs will actually be triggered when new entries are added to https://github.com/ros/rosdistro
1. There are a bunch of fixes in-progress on `rosdoc2` (our new tool for generating documentation), especially to support more complex C++ and Python projects (see https://github.com/ros-infrastructure/rosdoc2/pulls)
1. We need to actually get the destination URLs in place as laid out in https://design.ros2.org/articles/per_package_documentation.html
1. We need to develop buildfarm support for cross-linking between packages
1. We need to develop buildfarm support for documenting built portions of packages, most especially for generated messages and such

Of these, items 1, 2, and 3 are basically necessary before we can announce full support for this.  4 and 5 are nice-to-haves, but will make people's lives easier.

**Comment by clalancette** [comment]
Point 1 got completed.  We are reviewing PRs for point 2 above, hopefully that will get in relatively soon.  I've opened up https://github.com/ros2/ros2/issues/1209 to track point 3.  4 and 5 are nice-to-haves and will come later.

Given that, I'm going to close this issue out.  Feel free to open another issue at https://github.com/ros2/ros2 if you have more questions about this.
